src/freertos_drivers/esp32/Esp32WS2812.cxx

```diff
@@ -0,0 +1,311 @@
+/** \copyright
+ * Copyright (c) 2021, Mike Dunston
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file Esp32WS2812.cxx
+ *
+ * WS2812 LED strip controller using the RMT peripheral to generate the LED
+ * data stream.
+ *
+ * @author Mike Dunston
+ * @date 3 May 2021
+ */
+
+// Ensure we only compile this code for the ESP32
+#ifdef ESP32
+
+#include "freertos_drivers/esp32/Esp32WS2812.hxx"
+#include "utils/logging.h"
+#include <esp_idf_version.h>
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4,3,0)
+
+namespace openmrn_arduino
+{
+
+static uint32_t timeZeroHighTicks = 0;
+static uint32_t timeZeroLowTicks = 0;
+static uint32_t timeOneHighTicks = 0;
+static uint32_t timeOneLowTicks = 0;
+
+/// Number of microseconds to transmit a HIGH signal for a ZERO bit.
+static constexpr uint16_t TIME_ZERO_HIGH_NSEC = 350;
+
+/// Number of microseconds to transmit a LOW signal for a ZERO bit.
+static constexpr uint16_t TIME_ZERO_LOW_NSEC = 850;
+
+/// Number of microseconds to transmit a HIGH signal for a ONE bit.
+static constexpr uint16_t TIME_ONE_HIGH_NSEC = 800;
+
+/// Number of microseconds to transmit a LOW signal for a ONE bit.
+static constexpr uint16_t TIME_ONE_LOW_NSEC = 450;
+
+/// WS2812 to RMT data translation adapter for the ESP32 RMT peripheral.
+///
+/// @param src is the source data to be translated.
+/// @param dest is the RMT data buffer to translate into.
+/// @param src_size is the number of bytes in the source buffer.
+/// @param wanted_num is the number of bits to translate.
```

Comment:
Must be multiple of 8.

Reply:
agreed, this should be noted here and also in ESP-IDF. Will update

Reply:
Fixed